File: tools/compliance_checker.js
const fs = require('fs');
const path = require('path');

const REQUIRED_FILES = ["safecore.manifest.json"];
const BANNED_PATTERNS = ["AWS_ACCESS_KEY", "PRIVATE_KEY", "password="];

function checkManifest(targetPath) {
    const manifestPath = path.join(targetPath, "safecore.manifest.json");
    if (!fs.existsSync(manifestPath)) {
        console.error("❌ FAILED: safecore.manifest.json not found.");
        return false;
    }

    try {
        const data = JSON.parse(fs.readFileSync(manifestPath, 'utf8'));
        if (!data.compliance_level) {
            console.error("❌ FAILED: Manifest missing compliance_level.");
            return false;
        }
        console.log(`✅ Manifest found. Level: ${data.compliance_level}`);
        return true;
    } catch (e) {
        console.error(`❌ FAILED: Invalid JSON manifest. ${e.message}`);
        return false;
    }
}

function scanForSecrets(targetPath) {
    console.log("🔍 Scanning for banned secrets...");
    let violations = 0;

    function walkSync(dir, filelist) {
        const files = fs.readdirSync(dir);
        files.forEach(function (file) {
            const filePath = path.join(dir, file);
            if (fs.statSync(filePath).isDirectory()) {
                if (file !== 'node_modules' && file !== '.git' && file !== 'tools') {
                    walkSync(filePath, filelist);
                }
            } else {
                if (/\.(js|ts|py|java|md|json)$/.test(file)) {
                    // Check content
                    try {
                        const content = fs.readFileSync(filePath, 'utf8');
                        BANNED_PATTERNS.forEach(pattern => {
                            if (content.includes(pattern)) {
                                console.error(`⚠️ VIOLATION: '${pattern}' found in ${path.relative(targetPath, filePath)}`);
                                violations++;
                            }
                        });
                    } catch (err) {
                        // Ignore read errors
                    }
                }
            }
        });
    }

    walkSync(targetPath, []);

    if (violations > 0) {
        return false;
    }
    return true;
}

function main() {
    const targetPath = process.argv[2] || ".";
    console.log(`Starting SafeCore Compliance Scan on: ${targetPath}`);

    const passedManifest = checkManifest(targetPath);
    const passedSecrets = scanForSecrets(targetPath);

    if (passedManifest && passedSecrets) {
        console.log("\n✅ COMPLIANCE CHECK PASSED");
        process.exit(0);
    } else {
        console.log("\n❌ COMPLIANCE CHECK FAILED");
        process.exit(1);
    }
}

main();

File: tools/compliance_checker.py
import os
import json
import sys

REQUIRED_FILES = ["safecore.manifest.json"]
BANNED_PATTERNS = ["AWS_ACCESS_KEY", "PRIVATE_KEY", "password="]

def check_manifest(path):
    manifest_path = os.path.join(path, "safecore.manifest.json")
    if not os.path.exists(manifest_path):
        print("❌ FAILED: safecore.manifest.json not found.")
        return False
    
    try:
        with open(manifest_path, 'r') as f:
            data = json.load(f)
            if "compliance_level" not in data:
                 print("❌ FAILED: Manifest missing compliance_level.")
                 return False
            print(f"✅ Manifest found. Level: {data['compliance_level']}")
            return True
    except Exception as e:
        print(f"❌ FAILED: Invalid JSON manifest. {e}")
        return False

def scan_for_secrets(path):
    print("🔍 Scanning for banned secrets...")
    violations = 0
    for root, dirs, files in os.walk(path):
        for file in files:
            if file.endswith((".py", ".js", ".ts", ".java", ".md")):
                full_path = os.path.join(root, file)
                try:
                    with open(full_path, 'r', encoding='utf-8', errors='ignore') as f:
                        content = f.read()
                        for pattern in BANNED_PATTERNS:
                            if pattern in content:
                                print(f"⚠️ VIOLATION: '{pattern}' found in {file}")
                                violations += 1
                except:
                    pass
    if violations > 0:
        return False
    return True

def main():
    target_path = sys.argv[1] if len(sys.argv) > 1 else "."
    print(f"Starting SafeCore Compliance Scan on: {target_path}")
    
    passed_manifest = check_manifest(target_path)
    passed_secrets = scan_for_secrets(target_path)
    
    if passed_manifest and passed_secrets:
        print("\n✅ COMPLIANCE CHECK PASSED")
        sys.exit(0)
    else:
        print("\n❌ COMPLIANCE CHECK FAILED")
        sys.exit(1)

if __name__ == "__main__":
    main()
